File: data/src/main/java/com/tonitealive/app/data/net/RetrofitToniteAliveApi.java
package com.tonitealive.app.data.net;

import com.google.gson.GsonBuilder;
import com.tonitealive.app.data.TokenStore;
import com.tonitealive.app.domain.model.User;

import javax.inject.Inject;
import javax.inject.Singleton;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import rx.Observable;

@Singleton
public class RetrofitToniteAliveApi implements ToniteAliveApi {

    private final String baseUrl;
    private final TokenStore tokenStore;
    private final ApiService api;

    @Inject
    public RetrofitToniteAliveApi(String baseUrl, TokenStore tokenStore) {
        this.baseUrl = baseUrl;
        this.tokenStore = tokenStore;

        api = buildApiService(buildClient());
    }

    private ApiService buildApiService(OkHttpClient client) {
        GsonBuilder gsonBuilder = new GsonBuilder();
        Retrofit retrofit = new Retrofit.Builder()
                                .client(client)
                                .baseUrl(baseUrl)
                                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                                .addConverterFactory(GsonConverterFactory.create(gsonBuilder.create()))
                                .build();
        return retrofit.create(ApiService.class);
    }

    private OkHttpClient buildClient() {
        return new OkHttpClient.Builder()
                .addInterceptor(new ApiAuthInterceptor(tokenStore))
                .build();
    }

    @Override
    public Observable<User> getUserByUsername(String username) {
        return api.getUserByUsername(username);
    }

    @Override
    public Observable<User> getUserByEmail(String email) {
        return api.getUserByEmail(email);
    }
}

File: data/src/main/kotlin/com/tonitealive/app/data/net/RetrofitToniteAliveApi.kt
package com.tonitealive.app.data.net

import com.google.gson.GsonBuilder
import com.tonitealive.app.data.TokenStore
import com.tonitealive.app.domain.model.AuthToken
import com.tonitealive.app.domain.model.User
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import rx.Observable
import javax.inject.Singleton

@Singleton
class RetrofitToniteAliveApi(private val baseUrl: String,
                             private val tokenStore: TokenStore) : ToniteAliveApi {

    private val api = buildApiService(buildClient())

    private fun buildApiService(client: OkHttpClient): ApiService {
        // Register Gson modules
        val gsonBuilder = GsonBuilder()
        val retrofit = Retrofit.Builder()
                .client(client)
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create(gsonBuilder.create()))
                .build()
        return retrofit.create(ApiService::class.java)
    }

    private fun buildClient(): OkHttpClient {
        return OkHttpClient.Builder()
                    .addInterceptor(ApiAuthInterceptor(tokenStore))
                    .build()
    }

    override fun login(grantType: String, username: String, password: String): Observable<AuthToken> {
        return api.login(grantType, username, password)
    }

    override fun logout(): Observable<Void> {
        return api.logout()
    }

    override fun createUser(username: String, email: String, password: String): Observable<User> {
        return api.createUser(username, email, password)
    }

    override fun getUserByUsername(username: String): Observable<User> {
        return api.getUserByUsername(username)
    }

    override fun getUserByEmail(email: String): Observable<User> {
        return api.getUserByEmail(email)
    }

    override fun removeUser(username: String): Observable<Void> {
        return api.removeUser(username)
    }
}